server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -512,10 +512,10 @@ public void updateShardState(final ShardRouting newRouting,
                                 engine.rollTranslogGeneration();
                                 engine.fillSeqNoGaps(newPrimaryTerm);
                                 if (getMaxSeqNoOfUpdatesOrDeletes() == SequenceNumbers.UNASSIGNED_SEQ_NO) {
-                                    // TODO: Enable this assertion after we replicate max_seq_no_updates during replication
-                                    // assert indexSettings.getIndexVersionCreated().before(Version.V_7_0_0_alpha1) :
-                                    //    indexSettings.getIndexVersionCreated();
-                                    engine.initializeMaxSeqNoOfUpdatesOrDeletes();
+                                    // If the old primary was on an old version, this promoting primary (was replica before)
+                                    // does not have max_seq_no_of_updates. We need to bootstrap it manually from its local history.
+                                    assert indexSettings.getIndexVersionCreated().before(Version.V_7_0_0_alpha1);
+                                    engine.advanceMaxSeqNoOfUpdatesOrDeletes(seqNoStats().getMaxSeqNo());
```

Comment:
reads better thanks. I think this comment will be more intuitive if you change `old primary was on an old version` to `old primary was on an old version that didn't yet replicate the MSU, we need to bootstrap it... `